--- ghc-7.6.1/aclocal.m4.orig	2012-09-04 19:10:14.000000000 +0200
+++ ghc-7.6.1/aclocal.m4	2012-09-27 19:18:53.993056656 +0200
@@ -154,7 +154,7 @@
 [
     checkArch() {
         case [$]1 in
-        i386)
+        i386|i486|i586|i686)
             test -z "[$]2" || eval "[$]2=ArchX86"
             ;;
         x86_64)
@@ -185,7 +185,7 @@
 
     checkVendor() {
         case [$]1 in
-        dec|unknown|hp|apple|next|sun|sgi|ibm)
+        pld|dec|unknown|hp|apple|next|sun|sgi|ibm)
             ;;
         *)
             echo "Unknown vendor [$]1"
--- ghc-6.12.1/libraries/terminfo/configure.ac~	2009-09-15 08:09:05.000000000 +0200
+++ ghc-6.12.1/libraries/terminfo/configure.ac	2010-04-08 16:43:45.040705349 +0200
@@ -13,7 +13,8 @@
     [curses_libraries=NONE])
 
 
-TERMINFO_INCLUDE_DIRS=
+TERMINFO_INCLUDE_DIRS=/usr/include/ncursesw
+CPPFLAGS="-I/usr/include/ncursesw $CPPFLAGS"
 TERMINFO_LIB_DIRS=
 if test "x$curses_libraries" != "xNONE"; then
   LDFLAGS="-L$curses_libraries $LDFLAGS"
--- ghc-7.2.1/driver/split/ghc-split.lprl~	2011-08-07 19:10:05.000000000 +0200
+++ ghc-7.2.1/driver/split/ghc-split.lprl	2011-09-26 11:13:36.907176266 +0200
@@ -225,7 +225,7 @@
                             if $TargetPlatform =~ /-apple-darwin/m;
     return(&process_asm_block_m68k($str))  if $TargetPlatform =~ /^m68k-/m;
     return(&process_asm_block_sparc($str)) if $TargetPlatform =~ /^sparc-/m;
-    return(&process_asm_block_iX86($str))  if $TargetPlatform =~ /^i[34]86-/m;
+    return(&process_asm_block_iX86($str))  if $TargetPlatform =~ /^i[3456]86-/m;
     return(&process_asm_block_x86_64($str))  if $TargetPlatform =~ /^x86_64-/m;
     return(&process_asm_block_alpha($str)) if $TargetPlatform =~ /^alpha-/m;
     return(&process_asm_block_hppa($str))  if $TargetPlatform =~ /^hppa/m;
