--- ghc-9.6.1/m4/fptools_set_haskell_platform_vars.m4.orig	2023-03-11 06:42:58.000000000 +0100
+++ ghc-9.6.1/m4/fptools_set_haskell_platform_vars.m4	2023-05-05 13:10:17.735877257 +0200
@@ -5,7 +5,7 @@
 [
     checkArch() {
         case [$]1 in
-        i386)
+        i386|i486|i586|i686)
             test -z "[$]2" || eval "[$]2=ArchX86"
             ;;
         x86_64|amd64)
@@ -63,7 +63,7 @@
 
     checkVendor() {
         case [$]1 in
-        dec|none|unknown|hp|apple|next|sun|sgi|ibm|montavista|portbld|alpine)
+        dec|none|unknown|hp|apple|next|sun|sgi|ibm|montavista|portbld|alpine|pld)
             ;;
         *)
             AC_MSG_WARN([Unknown vendor [$]1])
--- ghc-9.6.1/utils/llvm-targets/gen-data-layout.sh.orig	2023-05-05 13:06:25.867871856 +0200
+++ ghc-9.6.1/utils/llvm-targets/gen-data-layout.sh	2023-05-05 13:10:09.979397765 +0200
@@ -55,6 +55,7 @@
     "aarch64-unknown-linux-gnu"
     "aarch64-unknown-linux-musl"
     "aarch64-unknown-linux"
+    "aarch64-pld-linux"
     # Linux x86
     "i386-unknown-linux-gnu"
     "i386-unknown-linux-musl"
@@ -65,9 +66,11 @@
     "i686-unknown-linux-gnu"
     "i686-unknown-linux-musl"
     "i686-unknown-linux"
+    "i686-pld-linux"
     "x86_64-unknown-linux-gnu"
     "x86_64-unknown-linux-musl"
     "x86_64-unknown-linux"
+    "x86_64-pld-linux"
     # Linux Android
     "x86_64-unknown-linux-android"
     "armv7-unknown-linux-androideabi"
