--- ghc-8.10.1/libraries/Cabal/Cabal/Distribution/Simple/InstallDirs.hs.orig	2019-12-30 03:32:20.000000000 +0100
+++ ghc-8.10.1/libraries/Cabal/Cabal/Distribution/Simple/InstallDirs.hs	2020-05-23 10:10:45.902028639 +0200
@@ -161,7 +161,7 @@
 -- users to be able to configure @--libdir=\/usr\/lib64@ for example but
 -- because by default we want to support installing multiple versions of
 -- packages and building the same package for multiple compilers we append the
--- libsubdir to get: @\/usr\/lib64\/$libname\/$compiler@.
+-- libsubdir to get: @\/usr\/lib64\/$compiler\/$libname@.
 --
 -- An additional complication is the need to support relocatable packages on
 -- systems which support such things, like Windows.
@@ -204,11 +204,11 @@
       libdir       = installLibDir,
       libsubdir    = case comp of
            UHC    -> "$pkgid"
-           _other -> "$abi" </> "$libname",
+           _other -> "$compiler" </> "$pkgid",
       dynlibdir    = "$libdir" </> case comp of
            UHC    -> "$pkgid"
-           _other -> "$abi",
-      libexecsubdir= "$abi" </> "$pkgid",
+           _other -> "$compiler" </> "$pkgid",
+      libexecsubdir= "$pkgid",
       flibdir      = "$libdir",
       libexecdir   = case buildOS of
         Windows   -> "$prefix" </> "$libname"
@@ -217,8 +217,8 @@
       datadir      = case buildOS of
         Windows   -> "$prefix"
         _other    -> "$prefix" </> "share",
-      datasubdir   = "$abi" </> "$pkgid",
-      docdir       = "$datadir" </> "doc" </> "$abi" </> "$pkgid",
+      datasubdir   = "$pkgid",
+      docdir       = "$datadir" </> "doc" </> "$pkgid",
       mandir       = "$datadir" </> "man",
       htmldir      = "$docdir"  </> "html",
       haddockdir   = "$htmldir",
