diff -urN ghc-6.2.org/configure.ac ghc-6.2/configure.ac
--- ghc-6.2.org/configure.ac	2004-06-10 03:10:14.698233240 +0200
+++ ghc-6.2/configure.ac	2004-06-10 03:11:07.826156568 +0200
@@ -99,7 +99,7 @@
         HostVendor_CPP='dec'
         HostOS_CPP='osf3'
         ;;
-alpha*-unknown-linux*)
+alpha*-*-linux*)
         HostPlatform=alpha-unknown-linux
         TargetPlatform=alpha-unknown-linux
         BuildPlatform=alpha-unknown-linux
@@ -443,7 +443,7 @@
         HostVendor_CPP='apple'
         HostOS_CPP='darwin'
         ;;
-powerpc-unknown-linux*)
+powerpc-*-linux*)
         HostPlatform=powerpc-unknown-linux
         TargetPlatform=powerpc-unknown-linux
         BuildPlatform=powerpc-unknown-linux

