--- ghc6-6.2.orig/mk/package.mk
+++ ghc6-6.2/mk/package.mk
@@ -237,10 +237,10 @@
 CLEAN_FILES += $(PACKAGE).haddock
 
 %.raw-hs : %.lhs
-	$(GHC) $(HC_OPTS) -D__HADDOCK__ -E -cpp $< -o $<.tmp && sed -e 's/^#.*//' <$<.tmp >$@
+	$(HC) $(HC_OPTS) -D__HADDOCK__ -E -cpp $< -o $<.tmp && sed -e 's/^#.*//' <$<.tmp >$@
 
 %.raw-hs : %.hs
-	$(GHC) $(HC_OPTS) -D__HADDOCK__ -E -cpp $< -o $<.tmp && sed -e 's/^#.*//' <$<.tmp >$@
+	$(HC) $(HC_OPTS) -D__HADDOCK__ -E -cpp $< -o $<.tmp && sed -e 's/^#.*//' <$<.tmp >$@
 
 install-docs :: $(HTML_DOC)
 	@$(INSTALL_DIR) $(datadir)/html/libraries/$(PACKAGE)
--- ghc6-6.2.orig/ghc/rts/Linker.c
+++ ghc6-6.2/ghc/rts/Linker.c
@@ -560,6 +560,7 @@
       Sym(__udivdi3)                            \
       Sym(__moddi3)                             \
       Sym(__umoddi3)				\
+      Sym(__muldi3)				\
       Sym(__ashldi3)				\
       Sym(__ashrdi3)				\
       Sym(__lshrdi3)				\
--- ghc6-6.2.orig/ghc/utils/prof/icons/Makefile
+++ ghc6-6.2/ghc/utils/prof/icons/Makefile
@@ -7,7 +7,7 @@
 TOP=../../..
 include $(TOP)/mk/boilerplate.mk
 
-override datadir=$(libdir)/icons
+override datadir:=$(datadir)/icons
 INSTALL_DATAS=$(wildcard *.xbm)
 
 include $(TOP)/mk/target.mk
--- ghc6-6.2.orig/ghc/utils/prof/Makefile
+++ ghc6-6.2/ghc/utils/prof/Makefile
@@ -29,7 +29,7 @@
 endif
 
 ifneq "$(BIN_DIST)" "1"
-SCRIPT_SUBST_VARS += libdir libexecdir
+SCRIPT_SUBST_VARS += datadir libdir libexecdir
 endif
 
 # don't recurse on 'make install'
--- ghc6-6.2.orig/ghc/utils/prof/ghcprof.prl
+++ ghc6-6.2/ghc/utils/prof/ghcprof.prl
@@ -20,7 +20,7 @@
 $bsp_g         = 13;
 $MaxNoNodes    = 1900;
 
-$icondir    = ( $INSTALLING ? "$libdir/icons" 
+$icondir    = ( $INSTALLING ? "$datadir/icons" 
 	                    : "$FPTOOLS_TOP_ABS/ghc/utils/prof/icons" );
  
 $xmlparser  = ( $INSTALLING ? "$libexecdir/xmlparser"
--- ghc6-6.2.orig/libraries/OpenGL/specs/enumerant/Makefile
+++ ghc6-6.2/libraries/OpenGL/specs/enumerant/Makefile
@@ -5,6 +5,7 @@
 
 # -----------------------------------------------------------------------------
 
+HC           = $(GHC_INPLACE)
 MKDEPENDHS   = $(GHC_INPLACE)
 HS_PROG      = ConvertEnumSpec
 CLEAN_FILES += Main.hi test.stdout.tmp
@@ -13,9 +14,9 @@
 
 ghc_ge_504 = $(shell if (test $(GhcCanonVersion) -ge 504); then echo YES; else echo NO; fi)
 
-ifeq "$(ghc_ge_504)" "NO"
-SRC_HC_OPTS += -package data -package text
-endif
+#ifeq "$(ghc_ge_504)" "NO"
+#SRC_HC_OPTS += -package data -package text
+#endif
 
 test: ConvertEnumSpec
 	$(RM) test.stdout.tmp
--- ghc6-6.2.orig/libraries/Makefile
+++ ghc6-6.2/libraries/Makefile
@@ -73,3 +73,9 @@
 	   echo $(INSTALL_DATA) $(INSTALL_OPTS) $$i $(datadir)/html/libraries; \
 	   $(INSTALL_DATA) $(INSTALL_OPTS) $$i $(datadir)/html/libraries; \
 	done
+
+CLEAN_FILES += config.status libraries.txt
+
+extraclean :: 
+	$(RM) -rf $(HTML_DIR)
+
--- ghc6-6.2.orig/configure
+++ ghc6-6.2/configure
@@ -1932,7 +1932,7 @@
 if test "$WithGhc" != ""; then
   echo "$as_me:$LINENO: checking version of ghc" >&5
 echo $ECHO_N "checking version of ghc... $ECHO_C" >&6
- "${WithGhc-ghc}" --version > conftestghc 2>&1
+ ${WithGhc-ghc} --version > conftestghc 2>&1
   cat conftestghc >&5
 #Useless Use Of cat award...
   fptools_version_of_ghc=`cat conftestghc | sed -n -e 's/, patchlevel *\([0-9]\)/.\1/;s/.* version \([0-9][0-9.]*\).*/\1/p'`
--- ghc6-6.2.orig/aclocal.m4
+++ ghc6-6.2/aclocal.m4
@@ -615,7 +615,7 @@
 dnl
 AC_DEFUN(FPTOOLS_GHC_VERSION,
 [FPTOOLS_NOCACHE_CHECK([version of ghc], [fptools_version_of_ghc],
-["${WithGhc-ghc}" --version > conftestghc 2>&1
+[${WithGhc-ghc} --version > conftestghc 2>&1
   cat conftestghc >&AC_FD_CC
 #Useless Use Of cat award...
   fptools_version_of_ghc=`cat conftestghc | sed -n -e 's/, patchlevel *\([[0-9]]\)/.\1/;s/.* version \([[0-9]][[0-9.]]*\).*/\1/p'`
--- ghc6-6.2.orig/debian/flags.xsl.in
+++ ghc6-6.2/debian/flags.xsl.in
@@ -0,0 +1,272 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
+<!DOCTYPE xsl:stylesheet [
+]>
+
+<xsl:stylesheet version="1.0"
+		xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+		xmlns="http://www.w3.org/TR/xhtml1/strict">
+				  
+<xsl:output method="text" omit-xml-declaration="yes" />
+
+<xsl:template match="/">.\"
+.\" This is a generated file.  Changes might get clobbered.  Edit at own's risk.
+.\"
+.TH GHC 1 "2002-10-25" "Glasgow FP Suite" "Glasgow Haskell Compiler"
+.SH NAME
+GHC \- the Glasgow Haskell Compiler
+
+
+.SH SYNOPSIS
+.B ghc
+.RI [ option | filename ]...
+.br
+.B ghci
+.RI [ option | filename ]...
+.br
+.B ghc6
+.RI [ option | filename ]...
+.br
+.B ghci6
+.RI [ option | filename ]...
+
+If you have the ghc6-threaded-rts package installed you will also have
+.B ghc6-threaded
+and
+.B ghci6-threaded
+variants.
+
+.SH DESCRIPTION
+This manual page documents briefly the
+.B ghc
+and
+.B ghci
+commands.
+Extensive documentation is available in various other formats
+including DVI, PostScript and HTML; see below.
+
+.PP
+Each of GHC's command line options is classified as either 
+.IR static " or " dynamic .
+A static flag may only be specified on the command line, whereas a
+dynamic flag may also be given in an \f(CROPTIONS\fP pragma in a
+source file or set from the GHCi command-line with \f(CR:set\fP.
+
+As a rule of thumb, all the language options are dynamic, as are the
+warning options and the debugging options.
+
+The rest are static, with the notable exceptions of
+.BR \-v ", " \-cpp ", " \-fasm ", " \-fvia\-C ", and " \-#include .
+The OPTIONS sections lists the status of each flag.
+
+.PP
+Common suffixes of file names for Haskell are:
+.TP
+.B .hs
+Haskell source code; preprocess, compile
+.TP
+.B .lhs
+literate Haskell source; unlit, preprocess, compile
+.TP
+.B .hi
+Interface file; contains information about exported
+symbols
+.TP
+.B .hc
+intermediate C files
+.TP
+.BI . x _o
+way 
+.I x 
+object files; common ways are:
+.BR p ", " u ", " s
+.TP
+.BI . x _hi
+way 
+.I x
+interface files
+
+
+.SH OPTIONS
+
+<xsl:apply-templates select="sect1/sect2" mode="overview"/>
+
+<xsl:apply-templates select="sect1/sect2"/>
+
+
+.SH FILES
+.I /usr/lib/ghc-@VERSION@/
+
+.SH COPYRIGHT
+
+Copyright 2002, The University Court of the University of Glasgow.
+.br
+All rights reserved.
+
+
+.SH AUTHOR
+
+This manual page was generated from the SGML documentation of GHC with blood,
+sweat, tears and a breaks-if-you-look-at-it-the-wrong-way XSL
+stylesheet by Michael Weber &lt;michaelw@debian.org&gt; for the Debian
+GNU/Linux system (but may be used by others).
+
+.\" End
+</xsl:template>
+
+
+<xsl:template match="sect1/sect2" mode="overview">
+<xsl:choose>
+<xsl:when test="contains(title/.,' (')">
+.SS <xsl:value-of select="substring-before(title/.,' (')"/>
+</xsl:when>
+<xsl:otherwise>
+.SS <xsl:value-of select="title/."/>
+</xsl:otherwise>
+</xsl:choose>
+.nh
+<xsl:apply-templates select="informaltable/tgroup/tbody/row" mode="overview"/>
+.hy
+</xsl:template>
+
+
+<xsl:template match="sect1/sect2">
+<xsl:choose>
+<xsl:when test="contains(title/.,' (')">
+
+.SH <xsl:value-of select='translate(substring-before(title/.," ("),"abcdefghijklmnopqrstuvwxyz","ABCDEFGHIJKLMNOPQRSTUVWXYZ")'/>
+
+</xsl:when>
+<xsl:otherwise>
+
+.SH <xsl:value-of select='translate(title/.,"abcdefghijklmnopqrstuvwxyz","ABCDEFGHIJKLMNOPQRSTUVWXYZ")'/>
+
+</xsl:otherwise>
+</xsl:choose><xsl:text>
+</xsl:text>
+<xsl:apply-templates select="informaltable/tgroup/tbody/row"/>
+</xsl:template>
+
+
+<xsl:template match="informaltable/tgroup/tbody/row" mode="overview">
+  <xsl:apply-templates select="entry[1]|entry[4]" mode="overview"/>
+  <xsl:text> </xsl:text>
+</xsl:template>
+
+<xsl:template match="informaltable/tgroup/tbody/row">
+.TP
+<xsl:apply-templates select="entry[1]"/><xsl:text>
+</xsl:text>
+<xsl:variable name="x">
+<xsl:apply-templates select="entry[2]"/>
+</xsl:variable>
+<xsl:value-of select="normalize-space($x)"/>
+.rj
+[<xsl:apply-templates select="entry[3]"/>]
+<!-- IGNORE NEGATIVE OPTIONS
+<xsl:if test="not(entry[4]='-')">
+  <xsl:text>.TP
+</xsl:text>
+  <xsl:apply-templates select="entry[4]/option"/>
+</xsl:if>
+ -->
+</xsl:template>
+
+
+<xsl:template match="option" mode="escape-dash">
+  <xsl:variable name="x">
+    <xsl:value-of select="."/>
+  </xsl:variable>
+  <xsl:variable name="y">
+    <xsl:call-template name="replace-string">
+      <xsl:with-param name="text" select="$x"/>
+      <xsl:with-param name="from" select="'-'"/>
+      <xsl:with-param name="to" select="'\-'"/>
+    </xsl:call-template>
+  </xsl:variable>
+  <xsl:value-of select="$y"/>
+</xsl:template>
+
+<xsl:template match="option" mode="overview">
+  <xsl:apply-templates select="." mode="escape-dash"/>
+</xsl:template>
+
+<xsl:template match="option">
+  <xsl:text>\fB</xsl:text>
+  <xsl:apply-templates select="." mode="escape-dash"/>
+  <xsl:text>\fP</xsl:text>
+</xsl:template>
+
+
+<xsl:template match="entry[1]" mode="overview">
+  <xsl:apply-templates mode="overview"/>
+  <xsl:text> </xsl:text>
+</xsl:template>
+
+<xsl:template match="entry[1]">
+  <xsl:apply-templates/><xsl:text> </xsl:text>
+</xsl:template>
+
+<xsl:template match="entry[4]" mode="overview">
+  <xsl:if test="not(.='-')">
+    <xsl:apply-templates select="option" mode="overview"/>
+    <xsl:text> </xsl:text>
+  </xsl:if>
+</xsl:template>
+
+<xsl:template match="entry[4]">
+  <xsl:if test="not(.='-')">
+    <xsl:value-of select="."/><xsl:text> </xsl:text>
+  </xsl:if>
+</xsl:template>
+
+
+<xsl:template match="replaceable" mode="overview">
+  <xsl:apply-templates select="."/>
+</xsl:template>
+
+<xsl:template match="replaceable">
+  <xsl:text>\fI</xsl:text>
+  <xsl:value-of select='.'/>
+  <xsl:text>\fP</xsl:text>
+</xsl:template>
+
+
+<xsl:template match="literal">
+  <xsl:text>\f(CR</xsl:text>
+  <xsl:value-of select="."/>
+  <xsl:text>\fP</xsl:text>
+</xsl:template>
+
+
+
+<!-- reusable replace-string function -->
+  <xsl:template name="replace-string">
+     <xsl:param name="text"/>
+     <xsl:param name="from"/>
+     <xsl:param name="to"/>
+
+     <xsl:choose>
+       <xsl:when test="contains($text, $from)">
+
+         <xsl:variable name="before" select="substring-before($text,
+	 $from)"/>
+         <xsl:variable name="after" select="substring-after($text,
+	 $from)"/>
+         <xsl:variable name="prefix" select="concat($before, $to)"/>
+
+         <xsl:value-of select="$before"/>
+         <xsl:value-of select="$to"/>
+         <xsl:call-template name="replace-string">
+           <xsl:with-param name="text" select="$after"/>
+           <xsl:with-param name="from" select="$from"/>
+           <xsl:with-param name="to" select="$to"/>
+         </xsl:call-template>
+       </xsl:when> 
+       <xsl:otherwise>
+         <xsl:value-of select="$text"/>  
+       </xsl:otherwise>
+     </xsl:choose>            
+  </xsl:template>
+
+
+</xsl:stylesheet>
--- ghc6-6.2.orig/debian/changelog
+++ ghc6-6.2/debian/changelog
@@ -0,0 +1,660 @@
+ghc6 (6.2-3) unstable; urgency=low
+
+  * More HOpenGL example tweaking.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sat, 24 Jan 2004 20:16:57 +0000
+
+ghc6 (6.2-2) unstable; urgency=low
+
+  * Apply fix so OpenGL docs build when the compiling compiler doesn't
+    support OpenGL.
+  * Tweaked HOpenGL example, including suggestions from Sven Panne (HOpenGL
+    author).
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Wed, 21 Jan 2004 23:57:22 +0000
+
+ghc6 (6.2-1) unstable; urgency=low
+
+  * New upstream version.
+  * ghc/docs/users_guide/flags.sgml is now xmlised by a script rather
+    than by hand.
+  * Added an HOpenGL example.
+  * Removed build dependency on happy (it has been pre-run on the tarball).
+  * Build dependency on haddock is now >= 0.6 (we need some of its new
+    options).
+  * For the normal ghc6 deb pass GhcLibsWithOpenGL=NO GhcLibsWithGLUT=NO
+    rather than GhcLibsWithHOpenGL=NO
+  * Add config.status and libraries.txt to CLEAN_FILES and a rule to
+    remove the HTML directory for the extraclean target in
+    libraries/Makefile
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Tue, 16 Dec 2003 18:09:11 +0000
+
+ghc6 (6.0.1-10) unstable; urgency=low
+
+  * Change #ifdef IN_STG_CODE to #if IN_STG_CODE in the alpha section of
+    ghc/includes/TailCalls.h
+  * Add __muldi3 to RTS_LIBGCC_SYMBOLS in ghc/rts/Linker.c (ghci was
+    complaining about it being unknown on sparc).
+  * Provide a /usr/bin/haskell-compiler alternative.
+  * Depend on haskell-utils
+  * Register /usr/bin/ghc6 and ghc6 with haskell-utils
+  * Register /usr/bin/ghc6-threaded and ghc6-threaded with haskell-utils
+  * Tidy up postinst/prerm.
+  * Add ghc6-threaded as an alternative for /usr/bin/ghc (priority 590,
+    i.e. between ghc5 and ghc6).
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sat, 11 Oct 2003 11:07:00 +0000
+
+ghc6 (6.0.1-9) unstable; urgency=low
+
+  * Move the TailCalls.h include in ghc/includes/Stg.h below the config.h
+    include. Fixes the undefined reference warnings in -8. Closes: #211430.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Wed, 17 Sep 2003 22:04:57 +0000
+
+ghc6 (6.0.1-8) unstable; urgency=low
+
+  * If we are not on an i386, sparc or ia64 machine then do an
+    unregisterised build.
+  * Unregisterised builds should be possible for any architecture, so
+    set the architecture to any.
+  * Add s390, m68k, mips, hppa, arm and powerpc Linux to configure{,.in}
+  * Write a README.Debian about the registerised/unregisterised deal.
+  * Change a couple of machine/foo.h includes on alpha to asm/foo.h
+  * Add the x86-64 hack from upstream to MBlock.h, generalising it to all
+    8-bit arches.
+  * Add -optc-mbig-switch to SRC_HC_OPTS in libraries/OpenGL/Makefile
+    on hppa (fixes an assembler failure for at least
+    Graphics/Rendering/OpenGL/GL/QueryUtils.p_o).
+  * Tweak ghc/includes/TailCalls.h so it only steals a register for STG
+    code.
+  * Move the include of TailCalls.h in ghc/includes/Stg.h up so it is
+    before all the procedure definitions.
+  * Change an IF_OS_darwin to an IF_ARCH_powerpc in
+    ghc/compiler/nativeGen/MachCode.lhs
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Tue, 16 Sep 2003 20:40:43 +0000
+
+ghc6 (6.0.1-7) unstable; urgency=low
+
+  * Add ia64 to Architecture field of the packages.
+  * Apply various IA64 fixes from upstream CVS.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Wed, 10 Sep 2003 13:47:56 +0000
+
+ghc6 (6.0.1-6) unstable; urgency=low
+
+  * We need docbook-utils even for a binary-only build as the build fails
+    if it can't build the documentation.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Mon, 08 Sep 2003 00:34:51 +0000
+
+ghc6 (6.0.1-5) unstable; urgency=low
+
+  * Add jade to the build-deps.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sun, 07 Sep 2003 10:11:54 +0000
+
+ghc6 (6.0.1-4) unstable; urgency=low
+
+  * Remove unnecessary flex build-dep and comment it out of
+    configure{,.in}.
+  * Resurrect the "SGMLDocWays  := html dvi ps" line in build.mk that
+    got lost.
+  * Do a test compilation+run after building to make sure things aren't
+    *completely* screwed up.
+  * make distclean rather than maintainer-clean
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sat, 06 Sep 2003 11:07:08 +0000
+
+ghc6 (6.0.1-3) unstable; urgency=low
+
+  * Put sparc-unknown-linux entry in configure and configure.in
+  * Add sparc Linux support to ghc/driver/mangler/ghc-asm.lprl
+  * Build depend on gcc-2.95 [sparc]
+  * Change ghc6's dependency on gcc to ${gcc}.
+    This is set to either gcc-2.95 (if we are on a sparc) or gcc
+    (otherwise) in debian/ghc6.substvars
+  * Removed quotes around ${WithGhc-ghc} in configure and aclocal.m4 so
+    arguments can be given (e.g. to point at a certain gcc) with the
+    flag
+  * Pass flags giving the location of ghc and, in the case of sparc,
+    gcc to ./configure
+  * Added sparc to Architecture field of the packages
+  * touch configure early in the build and clean targets to stop the
+    build system trying to regenerate it with autoconf.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Thu, 14 Aug 2003 07:24:18 -0400
+
+ghc6 (6.0.1-2) unstable; urgency=low
+
+  * Ask for the correct name to be removed by update-alternatives
+  * Have manpage variables substituted
+  * Fix problem with files not getting installed into ghc6-libsrc package
+  * Put ghc/{README,ANNOUNCE} in /usr/share/doc/ghc6
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sat, 02 Aug 2003 20:27:44 +0000
+
+ghc6 (6.0.1-1) unstable; urgency=low
+
+  * New upstream version
+  * Now policy 3.6.0 compliant
+  * Fixed manpage alternatives to create a manpage for ghc, not ghc6.
+  * Added manpages for ghci, ghci6 and the -threaded variants of
+    these and ghc. Manpage now also mentions ghci.
+  * Reinstate the -libsrc package.
+  * Remove redundant build dependencies on autoconf and autotools-dev
+  * Comment out fix for compiling GL stuff with older GHCs as it really
+    uses the in-place compiler
+  * Have all packages (except ghc6-doc due to issues with an empty depends
+    line) depend on ${shlibs:Depends}, ${misc:Depends}
+  * Create a /usr/bin/ghc6 and /usr/bin/ghci6
+  * Also build a threaded-rts binary and make various related -threaded files
+    and symlinks. Put in a separate package (ghc6-threaded-rts) due to size.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Wed, 30 Jul 2003 12:37:21 +0000
+
+ghc6 (6.0-2) unstable; urgency=low
+
+  * Fix GET_PROC_ADDRESS properly so opengl libraries can be used (from
+    upstream CVS)
+  * GLUT libraries are also installed in the hopengl deb
+  * Update build-depends and depends to libglut3-dev, xlibmesa-gl-dev
+    and xlibmesa-glu-dev from glutg3-dev and xlibmesa-dev.
+  * Require at least version 0.4.0 of haddock.
+  * Fix doc-base control file links.
+  * Policy 3.5.10 compliant.
+  * Create the hp2ps-ghc6.1 manpage correctly.  Closes: #199786.
+  * Change glActiveTexture to glActiveTextureARB. ghci now loads with
+    the OpenGL and GLUT packages.
+  * Update build-depend on debhelper to >= 4.
+  * Correct "GHC5" to "GHC6" in the doc-base title.
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sun, 06 Jul 2003 14:42:40 +0000
+
+ghc6 (6.0-1) unstable; urgency=low
+
+  * New upstream version
+  * New maintainer, with the old maintainer's blessing
+  * changed package name to ghc6
+  * Building process simplified as GHC's build system now handles multiple
+    stages itself
+  * Patches from upstream:
+    + Removed comments from cpp functions in a couple of source files to
+      counteract changes in 3.3 meaning they didn't compile
+    + Evil mangler tweaked to allow extra whitespace gcc 3.3 puts in
+      and for changes to generated assembly for _module_registered
+    + Rename "--list-packages-local" to "--list-local-packages"
+  * Added missing quotes to
+    libraries/OpenGL/Graphics/Rendering/OpenGL/GL/Extensions.hs
+    so it compiles
+  * Removed fptags as it doesn't have a clear licence; should probably
+    be in its own package anyway as I imagine it is equally useful with
+    any Haskell compiler
+  * Changed libraries/OpenGL/specs/enumerant to use this compiler rather
+    than the installed one; in particular this means that the profiling
+    libraries don't need to be installed
+
+ -- Ian Lynagh (wibble) <igloo@debian.org>  Sat, 07 Jun 2003 17:16:25 +0000
+
+ghc5 (5.04.2-1) unstable; urgency=low
+
+  * CVS Version from 2002-12-05 (tag: ghc-5-04-2)
+    + fix for inet_ntoa linking error (upstream)
+      (closes: Bug#164573)
+    + export TimeLocale constructor (upstream)
+      (closes: Bug#167647)
+  * implemented proper auto* handling (config.*, autoconf)
+  * added autotools-dev to build-deps, bumped needed autoconf version to
+    >= 2.52
+  * removed cpio from Build-Deps (no longer needed)
+  * added man page for ghc
+    (closes: Bug#110037)
+  * added links for user_guide/user-guide.html and hslibs/book-hslibs.html
+    (closes: Bug#153795)
+  * necessary adjustments for debian_libraries_hopengl.dpatch
+  * libdir adjustments (always in sync with Debian package version now)
+  * added notes in ghc5-hopengl.README.Debian
+  * removed cygnus-stylesheets from Build-Deps-Indep
+    (closes: Bug#170112)
+  * moved ghcprof icons to /usr/share/ tree
+    (debian_ghcprof_datadir.dpatch)
+  * edited debian/copyright to make lintian shut up ("Upstream Author(s)")
+  * backport of "*_stub.o link failure" fix from ghc-5-04-branch
+    (debian_compManager_findstubs.dpatch)
+    (closes: Bug#171518)
+
+ -- Michael Weber <michaelw@debian.org>  Sat,  7 Dec 2002 19:38:10 +0100
+
+ghc5 (5.04-1) unstable; urgency=low
+
+  * CVS Version from 2002-07-15 (tag: ghc-5-04)
+    (closes: Bug#153020)
+    + spurious ghci dynamic linkage error solved by upstream update
+      (closes: Bug#151126)
+  * added Depends: libreadline4-dev (though I don't like it
+    and it just covers a technical issue), because too many people
+    get bitten by this :(
+    (closes: Bug#134950)
+  * build process revamped
+    + much easier to understand/maintain (KISS)
+      (it's a build script, no need to be too fancy)
+    + support for full bootstrap builds (needed for buildds)
+    + moved source to upstream-ghc/
+    + adjusted patches; must contain `upstream-ghc/' in
+      the path from now on (new source layout)
+    + uses shadow trees (lndir)
+    + added "-e noidref" to docbook2* tool calls, so that they
+      don't barf on missing ids (there are 2 currently)
+    + use upstream doc install mechanism now that it's usable
+      for Debian (slight changes in the dir layout, though)
+    + removed support for `debian/rules cvs-update'
+    + removed ghc5-libsrc package, now that decent library
+      docs start to pop up
+    + removed debian_build.dpatch (now debian/build.mk-stage*)
+  * renamed ghc4_ghc_includes_gmp.dpatch debian_ghc_includes_gmp.dpatch
+    because it's Debian specific
+  * added Build-Depends: haddock
+  * adjusted *.doc-base.* again to reflect new doc layout
+    (only ghc5-doc.doc-base.users-guide left)
+  * _really_ changed CVS/Root to :pserver:anoncvs@cvs.haskell.org:/cvs
+    this time (no idea what happened to the last changes)
+  * updated CVS checkout instructions in copyright
+  * updated upstream authors in copyright
+  * updated README.Debian to reflect changes
+  * fixed missing "sec-concurrent-haskell" SGML id
+  * added hasktags program to update-alternatives
+  * new package ghc5-hopengl (finally!)
+    (closes: Bug#116984)
+  * added Build-Depends: xlibmesa-dev, glutg3-dev (for HOpenGL)
+
+ -- Michael Weber <michaelw@debian.org>  Fri, 19 Jul 2002 15:34:43 +0200
+
+ghc5 (5.02.2-1) unstable; urgency=low
+
+  * CVS Version from 2002-02-14 (tag: ghc-5-02-2)
+    (closes: Bug#133595)
+  * buggy Makefile which prevented building with make-3.77 fixed
+    by update
+    (closes: Bug#117883)
+  * The "You Asked For It" release =)
+  * _NOT_ the "I Solved All Problems" release
+  * changed CVS/Root to :pserver:anoncvs@cvs.haskell.org:/cvs
+  * updated copyright
+  * removed lib6-dev from Depends: (no longer necessary)
+  * docbook build-deps now: docbook-utils | cygnus-stylesheets
+    (closes: Bug#123180)
+  * added time to Depends: (lost during backporting potato stuff)
+  * added build-deps: xutils (for lndir), time (used by build process)
+  * added build-deps: devscripts (for debchange, thanks to
+    Peter Gammie)
+  * readline build-deps now: libreadline-dev | libreadline4-dev
+  * priority: extra -> optional (as requested from the ftpmasters)
+  * for the -prof package, section: libs -> devel (as requested from
+    the ftpmasters)
+  * spelling correction in control file (cpu -> CPU)
+    (closes: Bug#124668)
+  * removed getSocketOption patch
+    (ghc5_getSocketOption.dpatch)
+  * added index.html symlinks for html docs
+  * adjusted  *.doc-base.*
+  * removed ghc5-doc.doc-base.hslibs (accessible via 
+    the user's guide)
+  * bumped Standards-Version to 3.5.6.0
+
+ -- Michael Weber <michaelw@debian.org>  Thu, 14 Feb 2002 21:01:14 +0100
+
+ghc5 (5.02-1) unstable; urgency=low
+
+  * CVS Version from 2001-10-15 (tag: ghc-5-02-branch)
+  * adjusted debian_glafp-utils_docbook.dpatch
+  * typo in docs fixed
+    (closes: Bug#110038)
+  * added hotfix for getSocketOption b0rkage (already fixed in CVS)
+    (ghc5_getSocketOption.dpatch)
+
+ -- Michael Weber <michaelw@debian.org>  Mon, 15 Oct 2001 19:59:43 +0200
+
+ghc5 (5.00.2-1) unstable; urgency=low
+
+  * CVS Version from 2001-07-16 (tag: ghc-5-00-2)
+    (closes: Bug#106756)
+  * upstream fix for .ghci problem
+    (closes: Bug#94739)
+  * removed patches which went into upstream
+    (ghc5_glitches.dpatch)
+
+ -- Michael Weber <michaelw@debian.org>  Tue, 31 Jul 2001 10:54:45 +0200
+
+ghc5 (5.00-2) unstable; urgency=low
+
+  * added dependency on libc6-dev as a workaround
+    (ghci needs the libm.so link)
+  * adjusted ghc5_glitches.dpatch slightly
+
+ -- Michael Weber <michaelw@debian.org>  Thu, 26 Apr 2001 15:07:04 +0200
+
+ghc5 (5.00-1) unstable; urgency=low
+
+  * CVS Version from 2001-04-11 (tag: ghc-5-00)
+  * changed package name to ghc5
+  * some more reorganizations to prepare dual ghc4/ghc5 installation
+  * added some missing build-deps
+  * now using libgmp3(-dev)
+  * removed some obsolete patches
+    (ghc4_hclose_bug.dpatch,ghc4_signal_bug.dpatch)
+  * added some other patches
+    (ghc5_glitches.dpatch)
+  * updated build.mk-patch once more
+  * DH_COMPAT=2
+  * added more docs (ffi-art)
+  * added support for CVS happy's
+
+ -- Michael Weber <michaelw@debian.org>  Wed, 18 Apr 2001 00:19:35 +0200
+
+ghc4 (4.08.1-4) unstable; urgency=low
+
+  * first upload as proper maintainer
+  * changed email address to @debian.org
+
+ -- Michael Weber <michaelw@debian.org>  Sun, 14 Jan 2001 01:56:05 +0100
+
+ghc4 (4.08.1-3) unstable; urgency=low
+
+  * unofficial version
+  * backported signal-bug patch from CVS trunk
+    (ghc4_signal_bug.dpatch)
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Tue, 10 Oct 2000 13:34:28 +0200
+
+ghc4 (4.08.1-2) unstable; urgency=low
+
+  * unofficial version
+  * backported hClose-bug patch from CVS trunk
+    (ghc4_hclose_bug.dpatch)
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Tue, 10 Oct 2000 13:33:56 +0200
+
+ghc4 (4.08.1-1) unstable; urgency=low
+
+  * !!NOTE!!
+    This version replaced aaronv's upload of 4.04.19990916-3 (thanks
+    to wli@debian.org for the sponsor upload).  All versions in
+    between were never officially available on the debian servers.
+  * CVS Version from 2000-09-02 (4.08.1)
+  * official 4.08.1 version
+  * uncommented `cvs upd' commands (accidently commented)
+  * changed `cvs-update' target again (first step of upcoming
+    "experimental" support)
+  * updated build.mk-patch
+  * set.{ps,dvi} accidently were not copied into the package,
+    (fixed now)
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Mon, 25 Sep 2000 14:02:52 +0200
+
+ghc4 (4.08.0-1) unstable; urgency=low
+
+  * CVS Version from 2000-06-29 (4.08.0)
+  * this is a snap of the official 4.08 release candidate
+  * need (happy >= 1.7) from now on
+  * unset DOCBOOK_PREFIX in order to use installed db tools
+  * an installed ghc4-prof pkg. must have the same version as the
+    corresponding ghc4 pkg.
+  * added option `-k' to make when cleaning, so it won't stumble over
+    errors
+  * ghc now understands option '--numeric-version', which will be
+    supported in the upcoming unstable version anyway...
+    (ghc4_numeric_version.dpatch)
+  * sync'ed base documents in doc-base entries with reality
+  * rearranged `docs',`build' targets
+  * rearranged documentation to sync with upstream
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Mon,  3 Jul 2000 14:52:12 +0200
+
+ghc4 (4.07.20000526-1) unstable; urgency=low
+
+  * CVS Version from 2000-05-26 (4.07)
+  * update to ghc-4-07-branch
+  * disabled debian_glafp-utils_docbook.dpatch, because it's not
+    necessary on this branch
+  * !! UNRELEASED due to building problem !!
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Fri, 26 May 2000 20:33:32 +0200
+
+ghc4 (4.07.20000523-1) unstable; urgency=low
+
+  * CVS Version from 2000-05-23 (4.07)
+  * removed ghc4_docbook.dpatch (incorporated by upstream, finally :-))
+  * removed ghc4_hslibs_tools_clean.dpatch (incorporated by upstream)
+  * removed ghc4_hslibs_util_clean.dpatch (incorporated by upstream)
+  * removed building of docbook (Debian does have a working docbook
+    environment)
+    (debian_glafp-utils_docbook.dpatch)
+  * added fptags to distribution (in a really q'n'd way)
+  * added debian/rules options (UN)PATCHOPTS, default is now
+    `PATCHOPTS = --ignore-whitespace --forward'
+  * added notes about debian/rules options to README.Debian
+  * adjusted build depends
+  * added more hints in README.debian
+  * !! UNRELEASED due to building problem !!
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Fri, 26 May 2000 20:02:20 +0200
+
+ghc4 (4.07.20000504-1) unstable; urgency=low
+
+  * CVS Version from 2000-05-04 (4.07)
+  * hslibs/utils/Readline_stub.p_o not cleaned
+    (ghc4_hslibs_util_clean.dpatch)
+  * added DrIFT, DtdToHaskell, Xtract as update-alternatives
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Mon,  8 May 2000 01:05:37 +0200
+
+ghc4 (4.07.20000429-1) unstable; urgency=low
+
+  * CVS Version from 2000-04-29 (4.07)
+  * removed ghc4_hslibs_util_clean.dpatch (solved by upstream)
+  * Main.hi not cleaned in hslibs/tools/* dirs
+    (ghc4_hslibs_tools_clean.dpatch)
+  * changed clean targets from `realclean' to `maintainer-clean'
+    (due to upstream change)
+  * !! UNRELEASED due to stage1 problems !!
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sun, 30 Apr 2000 01:06:47 +0200
+
+ghc4 (4.07.20000411-1) unstable; urgency=low
+
+  * CVS Version from 2000-04-11 (4.07)
+  * removed ghc4_configure_readline.dpatch (incorporated by upstream)
+  * adjusted Build-depends (this version can only be build with GHC 4.06)
+  * `debian/rules patch' now doesn't try to apply patches, if they're
+    already applied
+  * added option `StripLibraries = YES'
+  * !! UNRELEASED due to stage2 problems !!
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Thu, 13 Apr 2000 00:37:29 +0200
+
+ghc4 (4.07.20000202-1) unstable; urgency=low
+
+  * CVS Version from 2000-02-02 (4.07)
+  * added target `cvs-update' (see README.Debian)
+  * replaced sgml-tools by cygnus-stylesheets in Build-Depends
+  * removed ghc4_hslibs_cstring.dpatch (incorporated by upstream)
+  * removed ghc4_ghc_mk.dpatch (incorporated by upstream)
+  * removed debian_sgml.dpatch (no longer necessary)
+  * use autoreconf instead of autoconf (in debian/rules.local)
+  * html docs back in action
+  * renamed doc-base ID ghc4-installation to ghc4-building
+  * YARP [Just Another Readline Patch]: doesn't find termcap
+    (ghc4_configure_readline.dpatch)
+  * Readline_stub--stamp was not removed on clean
+    (ghc4_hslibs_util_clean.dpatch)
+  * removed hardcoded DTD paths in docs/fptools-both.dsl
+    (ghc4_docbook.dpatch)
+  * added `docs' target
+  * `debian/rules clean' again checks for root priv. (cf. packaging#3.2.1)
+  * added build option `-split-objs' (produces smaller executables)
+  * added build option `GhcLibHcOpts = -O' (won't compile otherwise, because
+    of a compiler bug in 4.06)
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Thu,  3 Feb 2000 17:47:00 +0100
+
+ghc4 (4.06.20000109-1) unstable; urgency=low
+
+  * CVS version from 2000-01-09 (4.06)
+  * debian/rules clean doesn't check for root priv. any more
+  * removed emacs local variable (add-log-mailing-address) in changelog.
+    (closes: Bug#54515)
+  * references to `haskell-doc' in control and debian/*doc-base*.
+    (closes: Bug#52729)
+  * Arch back to `i386' only, although sparc should be possible, too (with
+    minor changes, really). (closes: Bug#52794)
+  * bumped Standards-Version to 3.1.0
+  * introduced build dependencies
+  * removed ghc4_clean.dpatch (incorporated by upstream)
+  * removed ghc4_makefiles.dpatch (incorporated by upstream)
+  * removed ghc4_mk.dpatch (incorporated by upstream)
+  * removed ghc4_readline.dpatch (incorporated by upstream)
+  * removed ghc4_stat2resid.dpatch (incorporated by upstream)
+  * autoconf changes to rules, rules.local
+  * corrected bug in ghc/mk/paths.mk (wrt. variable referencing)
+    (ghc4_ghc_mk.dpatch)
+  * corrected bug in hslibs/CString.lhs (ghc4_hslibs_cstring.dpatch)
+  * added patch for sgmltools, removed {html,txt} because of problems
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sun,  9 Jan 2000 15:45:36 +0100
+
+ghc4 (4.05.19991206-1) unstable; urgency=low
+
+  * CVS version from 1999-12-06 (4.05)
+  * removed distrib/Makefile-bin.in patch (solved by upstream)
+  * removed ghc/compiler/Makefile patch (solved by upstream)
+  * adjusted mk/target.mk patch
+  * moved $(libdir)/ghc/$(version) -> $(libdir)/ghc-$(version) (upstream-like)
+  * dropped users-guide.info (wasn't complete, anyway)
+  * added postscript and plain text docs
+  * heap size back to 120M :-(
+  * changed the build process heavily
+  * fixed `missing symbol in libHSutil*.a' (formerly libHSmisc*.a) bug
+  * some cleanup patches
+  * updated URLs to the Haskell report in control and debian/*doc-base*
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Wed,  8 Dec 1999 23:37:37 +0100
+
+ghc4 (4.04.19990916-3local1) unstable; urgency=low
+
+  * added warning message about fragile `clean' target
+  * cut down memory req's for building GHC (100M heap should be ok)
+  * Arch: field back to ``i386''
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Tue, 30 Nov 1999 23:32:36 +0100
+
+ghc4 (4.04.19990916-3) unstable; urgency=low
+
+  * First revision to be uploaded to Debian
+  * Changed maintainer fields where necessary
+  * Changed binary packages to Architecture: any
+
+ -- Aaron Van Couwenberghe <aaronv@debian.org>  Tue, 30 Nov 1999 14:27:18 -0800
+
+ghc4 (4.04.19990916-2) unstable; urgency=low
+
+  * fixed typo in copyright
+  * bumped Standards-Version to 3.0.1
+  * doc-base support
+  * moved (un)patch rules to debian/rules.local
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Wed, 17 Nov 1999 18:06:25 +0100
+
+ghc4 (4.04.19990916-1) unstable; urgency=low
+
+  * added haskell-doc, ghc4-prof to Suggests
+  * new version numbering scheme
+  * improved build process (targets stage1,clean)
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sun, 19 Sep 1999 21:49:25 +0200
+
+ghc4 (4.04-9) unstable; urgency=low
+
+  * added libgmp2-dev to Depends
+  * CVS version from 1999/09/16
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Thu, 16 Sep 1999 22:09:11 +0200
+
+ghc4 (4.04-8) unstable; urgency=low
+
+  * fixed manpage packaging bug in ghc4-prof
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sat,  7 Aug 1999 21:59:05 +0200
+
+ghc4 (4.04-7) unstable; urgency=low
+
+  * Standards-Version 3.0.0.0
+  * removed undocumented manpages (lintian complains)
+  * moved usr/info -> usr/share/info according to FHS
+  * added 'perl | perl5' to Depends
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sat,  7 Aug 1999 21:54:39 +0200
+
+ghc4 (4.04-6) unstable; urgency=low
+
+  * removed patch for happy version check (now in upstream)
+  * GHC is working again
+  * CVS version from 1999/08/03
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Wed,  4 Aug 1999 18:12:07 +0200
+
+ghc4 (4.04-5) unstable; urgency=low
+
+  * local version, GHC doesn't work
+  * removed hstags from alternatives because it's non-functional
+  * removed mkdependHS from alternatives because directly running it 
+    is deprecated now (now 'ghc -M')
+  * re-added info file
+  * removed obsolete HAVE_READLINE variable in mk/build.mk
+  * removed patch for {hscpp,mkdependHS} "absolute path" fix (now in upstream)
+  * moved to patch system (taken from glibc package)
+  * CVS version from 1999/07/31
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sun,  1 Aug 1999 02:09:44 +0200
+
+ghc4 (4.04-4) unstable; urgency=low
+
+  * got rid of absolute cpp paths in {hscpp,mkdependHS}
+  * removed info files due to problem with sgmltools
+  * first glibc-2.1.1 version
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Tue, 13 Jul 1999 15:35:28 +0000
+
+ghc4 (4.04-3) unstable; urgency=low
+
+  * moved profiling libs into separate package
+  * added hscpp as alternative (useful for hugs)
+  * CVS version from 1999/07/10
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sat, 10 Jul 1999 19:20:49 +0200
+
+ghc4 (4.04-2) unstable; urgency=low
+
+  * compiled with use of happy-1.6 parser generator
+  * fixed some "unclean" directories
+  * added info entry
+  * added dwww index-file
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sat,  3 Jul 1999 03:04:13 +0200
+
+ghc4 (4.04-1) unstable; urgency=low
+
+  * added missing *.prl files
+  * added mkdependHS as alternative
+  * new CVS version as of 1999/06/30
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Wed, 30 Jun 1999 19:53:09 +0200
+
+ghc4 (4.03-1) unstable; urgency=low
+
+  * Initial Release.
+
+ -- Michael Weber <michael.weber@post.rwth-aachen.de>  Sat, 26 Jun 1999 17:52:29 +0200
+
+
--- ghc6-6.2.orig/debian/copyright
+++ ghc6-6.2/debian/copyright
@@ -0,0 +1,48 @@
+This package was debianised by Ian Lynagh <igloo@debian.org> on
+Sat, 07 Jun 2003. It was based on Michael Weber <michaelw@debian.org>'s
+ghc5 package. All of the Debian packaging scripts written by us are
+released into the public domain.
+
+It was downloaded from
+http://www.haskell.org/ghc/dist/6.2/ghc-6.2-src.tar.bz2
+
+Upstream Authors: The GHC team
+    Simon Peyton Jones	<simonpj@microsoft.com>
+    Simon Marlow	<simonmar@microsoft.com>
+
+See http://www.haskell.org/ghc/contributors.html for more details.
+
+
+Copyright:
+
+The Glasgow Haskell Compiler License
+
+Copyright 2002, The University Court of the University of Glasgow.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+· Redistributions of source code must retain the above copyright notice,
+  this list of conditions and the following disclaimer.
+
+· Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+· Neither name of the University nor the names of its contributors may be
+  used to endorse or promote products derived from this software without
+  specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE UNIVERSITY COURT OF THE UNIVERSITY OF
+GLASGOW AND THE CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+UNIVERSITY COURT OF THE UNIVERSITY OF GLASGOW OR THE CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+DAMAGE.
--- ghc6-6.2.orig/debian/control
+++ ghc6-6.2/debian/control
@@ -0,0 +1,104 @@
+Source: ghc6
+Section: devel
+Priority: optional
+Maintainer: Ian Lynagh (wibble) <igloo@debian.org>
+Standards-Version: 3.6.0
+Build-Depends: debhelper (>= 4), libgmp3-dev, xsltproc,
+ libreadline-dev | libreadline4-dev, xlibmesa-gl-dev,
+ xlibmesa-glu-dev, libglut3-dev, time, xutils, devscripts,
+ haddock (>= 0.6), ghc6 | ghc5 | ghc4 (>= 4.08),
+ cpio, gcc-2.95 [sparc], jade, docbook-utils
+
+Package: ghc6
+Architecture: any
+Depends: haskell-utils, perl | perl5, ${gcc}, libgmp3-dev, libreadline4-dev, ${shlibs:Depends}, ${misc:Depends}
+Provides: haskell-compiler, ghc
+Conflicts: ghc4 (<= 4.08.1-4)
+Suggests: ghc6-prof, ghc6-doc, haskell-doc
+Description: GHC - the Glasgow Haskell Compilation system
+ Version 6 of the Glorious Glasgow Haskell Compilation system (GHC).  GHC is
+ a compiler for Haskell98.
+ .
+ Haskell is "the" standard lazy functional programming language. Haskell98
+ is the current version of the language. The language definition and
+ additional documentation can be found in the `haskell-doc' package.
+ Alternatively, there is an online version at
+ http://haskell.org/onlinereport/.
+
+Package: ghc6-threaded-rts
+Section: devel
+Architecture: any
+Depends: ghc6 (= ${Source-Version}), ${shlibs:Depends}, ${misc:Depends}
+Description: Threaded RTS for the Glasgow Haskell Compilation system
+ This package contains the GHC compiler with the threaded RTS enabled.
+ If you don't know what that means then you probably don't need this.
+ .
+ The commands ghc6-threaded and ghci6-threaded are equivalent to ghc
+ and ghci in the normal package.
+
+Package: ghc6-hopengl
+Section: devel
+Architecture: any
+Depends: ghc6 (= ${Source-Version}), xlibmesa-gl-dev, xlibmesa-glu-dev, libglut3-dev, ${shlibs:Depends}, ${misc:Depends}
+Provides: ghc-hopengl
+Suggests: ghc6-doc
+Description: HOpenGL libraries for the Glasgow Haskell Compilation system
+ This package contains additional HOpenGL libraries, including suitable
+ profiling libraries for the Glorious Glasgow Haskell Compilation
+ system (GHC).
+ .
+ The documentation for HOpenGL can be found in the ghc6-doc package. 
+
+Package: ghc6-prof
+Section: devel
+Architecture: any
+Provides: ghc-prof
+Depends: ghc6 (= ${Source-Version}), ${shlibs:Depends}, ${misc:Depends}
+Description: Profiling libraries for the Glasgow Haskell Compilation system
+ Version 6 of the Glorious Glasgow Haskell Compilation system (GHC).  GHC is
+ a compiler for Haskell98.
+ .
+ Haskell is "the" standard lazy functional programming language. Haskell98
+ is the current version of the language. The language definition and
+ additional documentation can be found in the `haskell-doc' package.
+ Alternatively, there is an online version at
+ http://haskell.org/onlinereport/.
+ .
+ This package contains additional profiling libraries. They are only needed,
+ if you want to take a closer look on where exactly your program burns CPU
+ cycles.
+
+Package: ghc6-doc
+Section: doc
+Architecture: all
+Suggests: haskell-doc
+Description: Documentation for the Glasgow Haskell Compilation system
+ Version 6 of the Glorious Glasgow Haskell Compilation system (GHC).  GHC is
+ a compiler for Haskell98.
+ .
+ Haskell is "the" standard lazy functional programming language. Haskell98
+ is the current version of the language. The language definition and
+ additional documentation can be found in the `haskell-doc' package.
+ Alternatively, there is an online version at
+ http://haskell.org/onlinereport/.
+ .
+ This package includes HTML, DVI and PS versions of the SGML-based
+ documentation around GHC.
+
+Package: ghc6-libsrc
+Section: doc
+Suggests: ghc6-doc, haskell-doc
+Architecture: all
+Description: Library Sources of GHC, the Glasgow Haskell Compilation system
+ Version 6 of the Glorious Glasgow Haskell Compilation system (GHC). GHC is
+ a compiler for Haskell98.
+ .
+ Haskell is "the" standard lazy functional programming language. Haskell98
+ is the current version of the language. The language definition and
+ additional documentation can be found in the `haskell-doc' package.
+ Alternatively, there is an online version at
+ http://haskell.org/onlinereport/.
+ .
+ This package contains the Haskell source code of the GHC libraries. It is
+ provided as additional documentation ("Use the source, Luke").
+
--- ghc6-6.2.orig/debian/rules
+++ ghc6-6.2/debian/rules
@@ -0,0 +1,200 @@
+#!/usr/bin/make -f
+# Sample debian/rules that uses debhelper.
+# This file is public domain software, originally written by Joey Hess.
+#
+# This version is for a multibinary package. It also allows you to build any
+# of the binary packages independantly, via binary-<package> targets.
+
+# Uncomment this to turn on verbose mode. 
+#export DH_VERBOSE=1
+
+# This has to be exported to make some magic below work.
+export DH_OPTIONS
+
+include ghc/mk/version.mk
+
+GHC=$(firstword $(shell bash -c "type -p ghc6 ghc5 ghc"))
+ifeq "$(shell dpkg-architecture -qDEB_HOST_ARCH)" "sparc"
+EXTRA_CONFIGURE_FLAGS=--with-gcc=/usr/bin/gcc-2.95 --with-ghc="$(GHC) -pgmc/usr/bin/gcc-2.95 -pgmP\"/usr/bin/gcc-2.95 -E -undef -traditional\""
+else
+EXTRA_CONFIGURE_FLAGS=--with-ghc="$(GHC)"
+endif
+
+build: build-stamp
+build-stamp:
+	dh_testdir
+
+	touch configure
+	chmod +x debian/xmlise-flags.sgml
+ifeq "$(shell dpkg-architecture -qDEB_HOST_ARCH)" "sparc"
+	echo "gcc=gcc-2.95" >> debian/ghc6.substvars
+else
+	echo "gcc=gcc" >> debian/ghc6.substvars
+endif
+
+	echo "SGMLDocWays   := html dvi ps" > mk/build.mk
+ifeq '$(findstring $(shell dpkg-architecture -qDEB_HOST_ARCH), i386 sparc ia64)' ''
+	echo "GhcUnregisterised=YES" >> mk/build.mk
+	echo "SplitObjs=NO" >> mk/build.mk
+endif
+	./configure --enable-hopengl --prefix=/usr --bindir='$${libdir}/bin' $(EXTRA_CONFIGURE_FLAGS)
+	$(MAKE)
+
+	mkdir debian/test-build
+	echo 'main = putStrLn "Foo"' > debian/test-build/foo.hs
+	ghc/compiler/ghc-inplace debian/test-build/foo.hs -o debian/test-build/foo
+	[ "$$(debian/test-build/foo)" = "Foo" ]                    
+	rm debian/test-build/*
+	echo 'main = putStrLn "Foo"' > debian/test-build/foo.hs
+	ghc/compiler/ghc-inplace debian/test-build/foo.hs -o debian/test-build/foo -O2
+	[ "$$(debian/test-build/foo)" = "Foo" ]
+
+	$(MAKE) -f debian/scripts.mk all
+	debian/xmlise-flags.sgml
+	xsltproc debian/flags.xsl ghc/docs/users_guide/flags.xml > debian/ghc6.1
+	for m in ghci6 {ghc6,ghci6}-threaded; do echo ".so man1/ghc6.1" > debian/$$m.1; done
+	cp ghc/utils/hp2ps/hp2ps.1 debian/hp2ps-ghc6.1
+
+	touch build-stamp
+
+	dh_installdirs
+
+	$(MAKE) prefix=`pwd`/debian/tmp/usr install GhcLibsWithOpenGL=NO GhcLibsWithGLUT=NO
+	$(MAKE) prefix=`pwd`/debian/ghc6-hopengl/usr -C libraries/OpenGL install
+	$(MAKE) prefix=`pwd`/debian/ghc6-hopengl/usr -C libraries/GLUT   install
+	$(MAKE) prefix=`pwd`/debian/ghc6-doc/usr datadir='$${prefix}/share/doc/ghc6-doc' install-docs
+	cd libraries && find . \( -name "*.hs" -o -name "*.lhs" \) -print0 | cpio -updV0 ../debian/ghc6-libsrc/usr/share/doc/ghc6-libsrc
+	cd hslibs    && find . \( -name "*.hs" -o -name "*.lhs" \) -print0 | cpio -updV0 ../debian/ghc6-libsrc/usr/share/doc/ghc6-libsrc
+
+	find debian/tmp -type f    -name "*.p_hi"    \
+	                       -o  -name "*_p.a"     \
+	                                        > debian/ghc6-prof.install
+	find debian/tmp \( -type f -o -type l \) !  -name "*.p_hi"    \
+	                                         !  -name "*_p.a"     \
+	                                        > debian/ghc6.install
+	dh_install
+	/usr/bin/install -c -m 755 -d debian/ghc6/usr/bin
+	ln -s /usr/lib/ghc-$(ProjectVersion)/bin/ghc  debian/ghc6/usr/bin/ghc6
+	ln -s /usr/lib/ghc-$(ProjectVersion)/bin/ghci debian/ghc6/usr/bin/ghci6
+	
+	# Threaded RTS stuff
+	cp ghc/compiler/main/SysTools.lhs ghc/compiler/main/SysTools.lhs.pre-threaded-rts
+	sed 's/"package\.conf"/"package-threaded.conf"/' < ghc/compiler/main/SysTools.lhs.pre-threaded-rts > ghc/compiler/main/SysTools.lhs
+
+	cp ghc/utils/ghc-pkg/Main.hs ghc/utils/ghc-pkg/Main.hs.pre-threaded-rts
+	sed 's@"/package\.conf"@"/package-threaded.conf"@' < ghc/utils/ghc-pkg/Main.hs.pre-threaded-rts > ghc/utils/ghc-pkg/Main.hs
+
+	cp ghc/driver/package.conf ghc/driver/package.conf.pre-threaded-rts
+	sed -e 's@"HSrts"@"HSrts-threaded"@' -e 's@extra_libraries = \["m", "gmp", "dl"\],@extra_libraries = \["m", "gmp", "dl", "pthread"\],@' < ghc/driver/package.conf.pre-threaded-rts > ghc/driver/package.conf
+
+	cp ghc/driver/package.conf.inplace ghc/driver/package.conf.inplace.pre-threaded-rts
+	sed 's@extra_libraries = \["m", "gmp", "dl"\],@extra_libraries = \["m", "gmp", "dl", "pthread"\],@' < ghc/driver/package.conf.inplace.pre-threaded-rts > ghc/driver/package.conf.inplace
+
+	./configure --enable-hopengl --prefix=/usr --bindir='$${libdir}/bin' --enable-threaded-rts $(EXTRA_CONFIGURE_FLAGS)
+	make
+
+	/usr/bin/install -c -m 755 -d debian/ghc6-threaded-rts/usr/bin
+	/usr/bin/install -c -m 755 -d debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/
+	/usr/bin/install -c -m 755 -s ghc/utils/ghc-pkg/ghc-pkg.bin debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/ghc-pkg-threaded.bin
+	/usr/bin/install -c -m 755 -s ghc/compiler/stage2/ghc-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/ghc-threaded-$(ProjectVersion)
+	/usr/bin/install -c -m 644    ghc/rts/libHSrts.a debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/libHSrts-threaded.a
+	/usr/bin/install -c -m 644    ghc/rts/HSrts.o debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/HSrts-threaded.o
+	/usr/bin/install -c -m 644    ghc/rts/libHSrts_p.a debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/libHSrts-threaded_p.a
+	/usr/bin/install -c -m 644    ghc/driver/package.conf debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/package-threaded.conf
+	sed "2s/-/-threaded-/2" < debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghc-$(ProjectVersion) > debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-threaded-$(ProjectVersion)
+	chmod --reference=debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghc-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-threaded-$(ProjectVersion)
+	sed "2s/-/-threaded-/2" < debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghci-$(ProjectVersion) > debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghci-threaded-$(ProjectVersion)
+	chmod --reference=debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghci-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghci-threaded-$(ProjectVersion)
+	sed "2s/\.bin/-threaded.bin/" < debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghc-pkg-$(ProjectVersion) > debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-pkg-threaded-$(ProjectVersion)
+	chmod --reference=debian/ghc6/usr/lib/ghc-$(ProjectVersion)/bin/ghc-pkg-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-pkg-threaded-$(ProjectVersion)
+	ln -s ghc-threaded-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-threaded
+	ln -s ghci-threaded-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghci-threaded
+	ln -s ghc-pkg-threaded-$(ProjectVersion) debian/ghc6-threaded-rts/usr/lib/ghc-$(ProjectVersion)/bin/ghc-pkg-threaded
+	ln -s /usr/lib/ghc-$(ProjectVersion)/bin/ghc-threaded debian/ghc6-threaded-rts/usr/bin/ghc6-threaded
+	ln -s /usr/lib/ghc-$(ProjectVersion)/bin/ghci-threaded debian/ghc6-threaded-rts/usr/bin/ghci6-threaded
+
+	mv ghc/compiler/main/SysTools.lhs.pre-threaded-rts ghc/compiler/main/SysTools.lhs
+	mv ghc/utils/ghc-pkg/Main.hs.pre-threaded-rts ghc/utils/ghc-pkg/Main.hs
+	mv ghc/driver/package.conf.pre-threaded-rts ghc/driver/package.conf
+	mv ghc/driver/package.conf.inplace.pre-threaded-rts ghc/driver/package.conf.inplace
+
+clean:
+	dh_testdir
+	dh_testroot
+	rm -f build-stamp
+
+	touch configure
+	# Tidy up in case the threaded-rts build broke
+	-mv ghc/compiler/main/SysTools.lhs.pre-threaded-rts ghc/compiler/main/SysTools.lhs
+	-mv ghc/utils/ghc-pkg/Main.hs.pre-threaded-rts ghc/utils/ghc-pkg/Main.hs
+	-mv ghc/driver/package.conf.pre-threaded-rts ghc/driver/package.conf
+	-mv ghc/driver/package.conf.inplace.pre-threaded-rts ghc/driver/package.conf.inplace
+	
+	-$(MAKE) distclean
+	find . -type l -print0 | xargs -0 rm -f
+	rm -f mk/config.mk mk/config.h mk/stamp-h
+	rm -f hslibs/doc/hslibs.out ghc/docs/users_guide/users_guide.out
+	rm -f ghc/driver/package.conf.inplace.old ghc/driver/package.conf.old
+	rm -f debian/ghc6{.install,-prof.install}
+	rm -f debian/ghc{,i}6{,-threaded}.1 debian/hp2ps-ghc6.1
+	rm -f ghc/docs/users_guide/flags.xml
+	rm -f mk/build.mk
+	rm -rf debian/test-build
+	$(MAKE) -f debian/scripts.mk clean
+
+	dh_clean
+
+# This single target is used to build all the packages, all at once, or
+# one at a time. So keep in mind: any options passed to commands here will
+# affect _all_ packages. Anything you want to only affect one package
+# should be put in another target, such as the install target.
+binary-common:
+	dh_testdir
+	dh_testroot
+	dh_installchangelogs
+	dh_installdocs
+	dh_installexamples
+	dh_installmenu
+#	dh_installdebconf
+#	dh_installlogrotate
+#	dh_installemacsen
+#	dh_installcatalogs
+#	dh_installpam
+#	dh_installmime
+#	dh_installinit
+	dh_installman
+#	dh_installcron
+#	dh_installinfo
+#	dh_undocumented
+	dh_strip
+	dh_link
+	dh_compress
+	# dh_compress doesn't work properly across packages so we hack it here
+	rm -f debian/ghc6-threaded-rts/usr/share/man/man1/ghc6-threaded.1
+	rm -f debian/ghc6-threaded-rts/usr/share/man/man1/ghci6-threaded.1
+	ln -sf ghc6.1.gz debian/ghc6-threaded-rts/usr/share/man/man1/ghc6-threaded.1.gz || true
+	ln -sf ghc6.1.gz debian/ghc6-threaded-rts/usr/share/man/man1/ghci6-threaded.1.gz || true
+	dh_fixperms
+#	dh_perl
+#	dh_python
+#	dh_makeshlibs
+	dh_installdeb
+	dh_shlibdeps
+	dh_gencontrol
+	dh_md5sums
+	dh_builddeb
+
+# Build architecture independant packages using the common target.
+binary-indep: build
+	$(MAKE) -f debian/rules DH_OPTIONS=-i binary-common
+
+# Build architecture dependant packages using the common target.
+binary-arch: build
+	$(MAKE) -f debian/rules DH_OPTIONS=-a binary-common
+
+# Any other binary targets build just one binary package at a time.
+binary-%: build
+	make -f debian/rules binary-common DH_OPTIONS=-p$*
+
+binary: binary-indep binary-arch
+.PHONY: build clean binary-indep binary-arch binary-common binary
--- ghc6-6.2.orig/debian/ghc6-threaded-rts.prerm.in
+++ ghc6-6.2/debian/ghc6-threaded-rts.prerm.in
@@ -0,0 +1,38 @@
+#! /bin/sh
+# prerm script for ghc6
+#
+# SOURCE: prerm.in
+
+set -e
+
+bindir=/usr/lib/ghc-@VERSION@/bin
+
+# summary of how this script can be called:
+#        * <prerm> `remove'
+#        * <old-prerm> `upgrade' <new-version>
+#        * <new-prerm> `failed-upgrade' <old-version>
+#        * <conflictor's-prerm> `remove' `in-favour' <package> <new-version>
+#        * <deconfigured's-prerm> `deconfigure' `in-favour'
+#          <package-being-installed> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see /usr/doc/packaging-manual/
+
+case "$1" in
+    remove|upgrade|deconfigure|failed-upgrade)
+        update-alternatives --remove ghc $bindir/ghc-threaded-@VERSION@
+        haskell-utils --remove-compiler /usr/bin/ghc6-threaded
+        haskell-utils --remove-compiler ghc6-threaded
+    ;;
+    *)
+        echo "prerm called with unknown argument \`$1'" >&2
+        exit 0
+    ;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
+
--- ghc6-6.2.orig/debian/ghc6.prerm.in
+++ ghc6-6.2/debian/ghc6.prerm.in
@@ -0,0 +1,40 @@
+#! /bin/sh
+# prerm script for ghc6
+#
+# SOURCE: prerm.in
+
+set -e
+
+execdir=/usr/bin
+bindir=/usr/lib/ghc-@VERSION@/bin
+
+# summary of how this script can be called:
+#        * <prerm> `remove'
+#        * <old-prerm> `upgrade' <new-version>
+#        * <new-prerm> `failed-upgrade' <old-version>
+#        * <conflictor's-prerm> `remove' `in-favour' <package> <new-version>
+#        * <deconfigured's-prerm> `deconfigure' `in-favour'
+#          <package-being-installed> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see /usr/doc/packaging-manual/
+
+case "$1" in
+    remove|upgrade|deconfigure|failed-upgrade)
+        update-alternatives --remove haskell-compiler $execdir/ghc6
+        update-alternatives --remove ghc $bindir/ghc-@VERSION@
+        haskell-utils --remove-compiler /usr/bin/ghc6
+        haskell-utils --remove-compiler ghc6
+    ;;
+    *)
+        echo "prerm called with unknown argument \`$1'" >&2
+        exit 0
+    ;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
+
--- ghc6-6.2.orig/debian/ghc6.README.Debian
+++ ghc6-6.2/debian/ghc6.README.Debian
@@ -0,0 +1,20 @@
+
+GHC can be compiled either "unregisterised" or "registerised". A
+registerised build is about 25% faster than an unregisterised build, but
+it requires some bits of code be written specially for the architecture
+in question. Currently Debian has registerised builds for i386, sparc
+and ia64.
+
+Part of the work has been done for powerpc, m68k, alpha, mips (=> mipsel
+too?) and hppa. This means that either it works but for a different
+operating system or that it worked at some point in the past but hasn't
+been maintained. No attempt has been made for arm or s390. Interested
+experts on any of these architectures are encouraged to try to get GHC
+working registerised; some information on what is needed is available at
+http://www.haskell.org/ghc/docs/latest/html/building/sec-porting-ghc.html
+and you may be able to get help on the GHC mailing lists if you get
+stuck.
+
+
+Ian Lynagh <igloo@debian.org>
+
--- ghc6-6.2.orig/debian/ghc6.docs
+++ ghc6-6.2/debian/ghc6.docs
@@ -0,0 +1,3 @@
+ghc/README
+ghc/ANNOUNCE
+
--- ghc6-6.2.orig/debian/ghc6.postinst.in
+++ ghc6-6.2/debian/ghc6.postinst.in
@@ -0,0 +1,59 @@
+#! /bin/sh
+# postinst script for ghc6
+#
+# SOURCE: postinst.in
+set -e
+
+execdir=/usr/bin
+bindir=/usr/lib/ghc-@VERSION@/bin
+mandir=/usr/share/man
+
+# summary of how this script can be called:
+#        * <postinst> `configure' <most-recently-configured-version>
+#        * <old-postinst> `abort-upgrade' <new version>
+#        * <conflictor's-postinst> `abort-remove' `in-favour' <package>
+#          <new-version>
+#        * <deconfigured's-postinst> `abort-deconfigure' `in-favour'
+#          <failed-install-package> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see /usr/doc/packaging-manual/
+#
+# quoting from the policy:
+#     Any necessary prompting should almost always be confined to the
+#     post-installation script, and should be protected with a conditional
+#     so that unnecessary prompting doesn't happen if a package's
+#     installation fails and the `postinst' is called with `abort-upgrade',
+#     `abort-remove' or `abort-deconfigure'.
+
+case "$1" in
+    configure|abort-upgrade|abort-remove|abort-deconfigure)
+        update-alternatives \
+            --install $execdir/haskell-compiler haskell-compiler $execdir/ghc6 600 \
+            --slave $mandir/man1/haskell-compiler.1.gz haskell-compiler.1.gz $mandir/man1/ghc6.1.gz
+        update-alternatives \
+            --install $execdir/ghc ghc $bindir/ghc-@VERSION@ 600 \
+            --slave $execdir/ghc-pkg ghc-pkg $bindir/ghc-pkg-@VERSION@ \
+            --slave $execdir/ghcprof ghcprof $bindir/ghcprof \
+            --slave $execdir/ghci ghci $bindir/ghci-@VERSION@         \
+            --slave $execdir/hsc2hs hsc2hs $bindir/hsc2hs     \
+            --slave $execdir/hp2ps hp2ps $bindir/hp2ps    \
+            --slave $mandir/man1/hp2ps.1.gz hp2ps.1.gz $mandir/man1/hp2ps-ghc6.1.gz \
+            --slave $mandir/man1/ghc.1.gz ghc.1.gz $mandir/man1/ghc6.1.gz \
+            --slave $mandir/man1/ghci.1.gz ghci.1.gz $mandir/man1/ghc6.1.gz \
+            --slave $execdir/stat2resid stat2resid $bindir/stat2resid \
+            --slave $execdir/hasktags hasktags $bindir/hasktags
+        haskell-utils --add-compiler /usr/bin/ghc6 GHC @VERSION@
+        haskell-utils --add-compiler ghc6 GHC @VERSION@
+    ;;
+    *)
+        echo "postinst called with unknown argument \`$1'" >&2
+        exit 0
+    ;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
--- ghc6-6.2.orig/debian/compat
+++ ghc6-6.2/debian/compat
@@ -0,0 +1 @@
+4
--- ghc6-6.2.orig/debian/scripts.mk
+++ ghc6-6.2/debian/scripts.mk
@@ -0,0 +1,18 @@
+#!/usr/bin/make -f
+
+include ghc/mk/version.mk
+
+FILES  = debian/ghc6.postinst              debian/ghc6.prerm
+FILES += debian/ghc6-threaded-rts.postinst debian/ghc6-threaded-rts.prerm
+FILES += debian/flags.xsl
+
+.PHONY: all clean
+
+all: $(FILES)
+
+%: %.in
+	sed "s/@VERSION@/$(ProjectVersion)/" $< > $@
+
+clean:
+	rm -f $(FILES)
+
--- ghc6-6.2.orig/debian/ghc6-threaded-rts.postinst.in
+++ ghc6-6.2/debian/ghc6-threaded-rts.postinst.in
@@ -0,0 +1,56 @@
+#! /bin/sh
+# postinst script for ghc6-threaded-rts
+#
+
+set -e
+
+execdir=/usr/bin
+bindir=/usr/lib/ghc-@VERSION@/bin
+mandir=/usr/share/man
+
+# summary of how this script can be called:
+#        * <postinst> `configure' <most-recently-configured-version>
+#        * <old-postinst> `abort-upgrade' <new version>
+#        * <conflictor's-postinst> `abort-remove' `in-favour' <package>
+#          <new-version>
+#        * <deconfigured's-postinst> `abort-deconfigure' `in-favour'
+#          <failed-install-package> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see /usr/doc/packaging-manual/
+#
+# quoting from the policy:
+#     Any necessary prompting should almost always be confined to the
+#     post-installation script, and should be protected with a conditional
+#     so that unnecessary prompting doesn't happen if a package's
+#     installation fails and the `postinst' is called with `abort-upgrade',
+#     `abort-remove' or `abort-deconfigure'.
+
+case "$1" in
+    configure|abort-upgrade|abort-remove|abort-deconfigure)
+        update-alternatives \
+            --install $execdir/ghc ghc $bindir/ghc-threaded-@VERSION@ 590 \
+            --slave $execdir/ghc-pkg ghc-pkg $bindir/ghc-pkg-threaded-@VERSION@ \
+            --slave $execdir/ghcprof ghcprof $bindir/ghcprof \
+            --slave $execdir/ghci ghci $bindir/ghci-threaded-@VERSION@         \
+            --slave $execdir/hsc2hs hsc2hs $bindir/hsc2hs     \
+            --slave $execdir/hp2ps hp2ps $bindir/hp2ps    \
+            --slave $mandir/man1/hp2ps.1.gz hp2ps.1.gz $mandir/man1/hp2ps-ghc6.1.gz \
+            --slave $mandir/man1/ghc.1.gz ghc.1.gz $mandir/man1/ghc6-threaded.1.gz \
+            --slave $mandir/man1/ghci.1.gz ghci.1.gz $mandir/man1/ghc6-threaded.1.gz \
+            --slave $execdir/stat2resid stat2resid $bindir/stat2resid \
+            --slave $execdir/hasktags hasktags $bindir/hasktags
+        haskell-utils --add-compiler /usr/bin/ghc6-threaded GHC @VERSION@
+        haskell-utils --add-compiler ghc6-threaded GHC @VERSION@
+    ;;
+    *)
+        echo "postinst called with unknown argument \`$1'" >&2
+        exit 0
+    ;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
--- ghc6-6.2.orig/debian/ghc6-doc.doc-base.users-guide
+++ ghc6-6.2/debian/ghc6-doc.doc-base.users-guide
@@ -0,0 +1,22 @@
+Document: ghc6-users-guide
+Title: GHC6 User's manual
+Author: The GHC Team
+Abstract: Version 6 of the Glorious Glasgow Haskell Compilation system
+ (GHC). GHC is a compiler for Haskell98.
+ <br>
+ Haskell is "the" standard lazy functional programming language. Haskell98
+ is the current version of the language. The language definition and
+ additional documentation can be found in the `haskell-doc' package.
+ Alternatively, there is an online version at
+ http://haskell.org/onlinereport/.
+Section: devel
+
+Format: ps
+Files: /usr/share/doc/ghc6-doc/users_guide.ps.gz
+
+Format: dvi
+Files: /usr/share/doc/ghc6-doc/users_guide.dvi.gz
+
+Format: HTML
+Index: /usr/share/doc/ghc6-doc/html/index.html
+Files: /usr/share/doc/ghc6-doc/html/*/*.html
--- ghc6-6.2.orig/debian/ghc6-hopengl.README.debian
+++ ghc6-6.2/debian/ghc6-hopengl.README.debian
@@ -0,0 +1,17 @@
+
+HOpenGL was built as a part of the GHC compiler.  Note that this is
+-not- the 'official' (aka. standalone) version of HOpenGL.  Due to the
+hierarchical library transition, not everything is working yet.  See
+<URL:http://bugs.debian.org/156339:> for the full story.
+
+
+Documentation for the HOpenGL libraries can be found in the ghc6-doc
+package.
+
+Additional material, examples, screenshots, tutorials etc. are
+available online:
+
+http://www.haskell.org/HOpenGL/
+http://www.cin.ufpe.br/~haskell/hopengl/
+
+
--- ghc6-6.2.orig/debian/ghc6-threaded-rts.manpages
+++ ghc6-6.2/debian/ghc6-threaded-rts.manpages
@@ -0,0 +1,3 @@
+debian/ghc6-threaded.1
+debian/ghci6-threaded.1
+
--- ghc6-6.2.orig/debian/ghc6.manpages
+++ ghc6-6.2/debian/ghc6.manpages
@@ -0,0 +1,4 @@
+debian/ghc6.1
+debian/ghci6.1
+debian/hp2ps-ghc6.1
+
--- ghc6-6.2.orig/debian/xmlise-flags.sgml
+++ ghc6-6.2/debian/xmlise-flags.sgml
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+(
+cat <<'EOF'
+<!DOCTYPE sect1 [
+<!ENTITY ndash  "-">
+<!ENTITY ldquo  "`">
+<!ENTITY rdquo  "'">
+]>
+EOF
+#sed -e 's/& /\&amp; /g' \
+#    -e 's/<xref \([^>]*\)>/<xref \1 \/>/g' \
+#    -e 's/=\([^" >]\+\)/="\1"/g' \
+perl -pe 's/& /\&amp; /g;
+          s/<xref ([^>]*)>/<xref \1 \/>/g;
+          s/=([^" <>]+)(?=[^<]*>)/="\1"/g' \
+  ghc/docs/users_guide/flags.sgml
+) > ghc/docs/users_guide/flags.xml
+
--- ghc6-6.2.orig/debian/examples/HOpenGL/Cube.hs
+++ ghc6-6.2/debian/examples/HOpenGL/Cube.hs
@@ -0,0 +1,116 @@
+
+-- Written by Ian Lynagh <igloo@debian.org> in 2003.
+-- Released into the public domain.
+
+module Main (main) where
+
+import Graphics.UI.GLUT
+import System(ExitCode(..), exitWith)
+import System.Time (getClockTime, ClockTime(..))
+import Data.IORef
+
+type St = (ClockTime, GLfloat, GLfloat, GLfloat)
+
+dx, dy, dz :: GLfloat
+dx = 0.02
+dy = 0.05
+dz = 0.09
+
+red, green, yellow, blue, purple, cyan :: Color3 GLfloat
+red    = Color3 1 0 0
+green  = Color3 0 1 0
+yellow = Color3 1 1 0
+blue   = Color3 0 0 1
+purple = Color3 1 0 1
+cyan   = Color3 0 1 1
+
+main :: IO ()
+main = do getArgsAndInitialize
+          initialDisplayMode $= [ WithDepthBuffer, DoubleBuffered, RGBAMode ]
+          _win <- createWindow "Hello World"
+          myInit
+          depthFunc $= Just Less
+          drawBuffer $= BackBuffers
+          t <- getClockTime
+          ref <- newIORef (t, 0, 0, 0)
+          displayCallback $= (display ref)
+          keyboardMouseCallback $= Just keyboardMouse
+          idleCallback $= Just (inc_anim ref)
+          mainLoop
+
+inc_anim :: IORef St -> IO ()
+inc_anim x = do (t, r_x, r_y, r_z) <- readIORef x
+                t' <- getClockTime
+                let d = fromIntegral $ getMilliSecDiff t' t
+                writeIORef x (t', r_x + d*dx, r_y + d*dy, r_z + d*dz)
+                draw_colourful_cube x
+
+myInit :: IO () 
+myInit = do clearColor $= (Color4 0.0 0.0 0.0 0.0)
+            matrixMode $= Projection
+            loadIdentity
+            -- ortho 0.0 1.0 0.0 1.0 (-1.0) 1.0
+            -- frustum left right bottom top near far
+            frustum (-1) 1 (-1) 1 (-1) 40
+            matrixMode $= Modelview 0
+
+keyboardMouse :: KeyboardMouseCallback
+keyboardMouse (Char '\27') _ _ _ = exitWith ExitSuccess
+keyboardMouse (Char 'q')   _ _ _ = exitWith ExitSuccess
+keyboardMouse _ _ _ _ = return ()
+
+display :: IORef St -> DisplayCallback
+display x = draw_colourful_cube x
+
+draw_colourful_cube :: IORef St -> IO ()
+draw_colourful_cube x =
+               do (_, r_x, r_y, r_z) <- readIORef x
+                  clear [DepthBuffer, ColorBuffer]
+                  loadIdentity
+                  rotate r_x (Vector3 1 0 0 :: Vector3 GLfloat)
+                  rotate r_y (Vector3 0 1 0 :: Vector3 GLfloat)
+                  rotate r_z (Vector3 0 0 1 :: Vector3 GLfloat)
+                  mapM_ draw_face (zip colours faces)
+                  swapBuffers
+    where draw_face :: (Color3 GLfloat, IO ()) -> IO ()
+          draw_face (colour, face) = do color colour
+                                        renderPrimitive Quads face
+          faces = map (mapM_ vertex) face_vertices :: [IO ()]
+          colours = [red, green, yellow, blue, purple, cyan]
+          face_vertices = [
+                 [Vertex3     to     to     to,
+                  Vertex3   from     to     to,
+                  Vertex3   from   from     to,
+                  Vertex3     to   from     to],
+                 [Vertex3     to     to   from,
+                  Vertex3   from     to   from,
+                  Vertex3   from   from   from,
+                  Vertex3     to   from   from],
+                 [Vertex3     to     to     to,
+                  Vertex3   from     to     to,
+                  Vertex3   from     to   from,
+                  Vertex3     to     to   from],
+                 [Vertex3     to   from     to,
+                  Vertex3   from   from     to,
+                  Vertex3   from   from   from,
+                  Vertex3     to   from   from],
+                 [Vertex3     to     to     to,
+                  Vertex3     to   from     to,
+                  Vertex3     to   from   from,
+                  Vertex3     to     to   from],
+                 [Vertex3   from     to     to,
+                  Vertex3   from   from     to,
+                  Vertex3   from   from   from,
+                  Vertex3   from     to   from]]
+
+from, to :: GLfloat
+from = -0.4
+to   =  0.4
+
+getMilliSecDiff :: ClockTime -> ClockTime -> Integer
+getMilliSecDiff (TOD s1 u1) (TOD s2 u2) =
+    let d = (s1 - s2) * sec + u1 `div` to_milli - u2 `div` to_milli
+    in if d >= 0 then d else error "Time going backwards"
+  where sec = 10^(3 :: Int)
+        to_milli = 10^(9 :: Int)
+
--- ghc6-6.2.orig/debian/examples/HOpenGL/Makefile
+++ ghc6-6.2/debian/examples/HOpenGL/Makefile
@@ -0,0 +1,7 @@
+
+all:
+	ghc6 -Wall --make Cube -o cube
+
+clean:
+	rm -f *.o *.hi cube
+
--- ghc6-6.2.orig/debian/ghc6-hopengl.examples
+++ ghc6-6.2/debian/ghc6-hopengl.examples
@@ -0,0 +1,3 @@
+debian/examples/HOpenGL/Makefile
+debian/examples/HOpenGL/Cube.hs
+
