diff -urN ghc-6.2.1.org/aclocal.m4 ghc-6.2.1/aclocal.m4
--- ghc-6.2.1.org/aclocal.m4	2004-06-10 11:22:06.173807016 +0200
+++ ghc-6.2.1/aclocal.m4	2004-06-10 11:22:22.514322880 +0200
@@ -620,7 +620,7 @@
 dnl
 AC_DEFUN(FPTOOLS_GHC_VERSION,
 [FPTOOLS_NOCACHE_CHECK([version of ghc], [fptools_version_of_ghc],
-["${WithGhc-ghc}" --version > conftestghc 2>&1
+[${WithGhc-ghc} --version > conftestghc 2>&1
   cat conftestghc >&AC_FD_CC
 #Useless Use Of cat award...
   fptools_version_of_ghc=`cat conftestghc | sed -n -e 's/, patchlevel *\([[0-9]]\)/.\1/;s/.* version \([[0-9]][[0-9.]]*\).*/\1/p'`
diff -urN ghc-6.2.1.org/configure ghc-6.2.1/configure
--- ghc-6.2.1.org/configure	2004-06-10 11:22:06.172807168 +0200
+++ ghc-6.2.1/configure	2004-06-10 11:22:22.509323640 +0200
@@ -1937,7 +1937,7 @@
 if test "$WithGhc" != ""; then
   echo "$as_me:$LINENO: checking version of ghc" >&5
 echo $ECHO_N "checking version of ghc... $ECHO_C" >&6
- "${WithGhc-ghc}" --version > conftestghc 2>&1
+ ${WithGhc-ghc} --version > conftestghc 2>&1
   cat conftestghc >&5
 #Useless Use Of cat award...
   fptools_version_of_ghc=`cat conftestghc | sed -n -e 's/, patchlevel *\([0-9]\)/.\1/;s/.* version \([0-9][0-9.]*\).*/\1/p'`
diff -urN ghc-6.2.1.org/ghc/rts/Linker.c ghc-6.2.1/ghc/rts/Linker.c
--- ghc-6.2.1.org/ghc/rts/Linker.c	2004-06-10 11:22:04.674035016 +0200
+++ ghc-6.2.1/ghc/rts/Linker.c	2004-06-10 11:22:22.459331240 +0200
@@ -574,6 +574,7 @@
       Sym(__udivdi3)                            \
       Sym(__moddi3)                             \
       Sym(__umoddi3)				\
+      Sym(__muldi3)				\
       Sym(__ashldi3)				\
       Sym(__ashrdi3)				\
       Sym(__lshrdi3)				\
diff -urN ghc-6.2.1.org/ghc/utils/prof/ghcprof.prl ghc-6.2.1/ghc/utils/prof/ghcprof.prl
--- ghc-6.2.1.org/ghc/utils/prof/ghcprof.prl	2004-06-10 11:22:04.805015104 +0200
+++ ghc-6.2.1/ghc/utils/prof/ghcprof.prl	2004-06-10 11:22:22.464330480 +0200
@@ -20,7 +20,7 @@
 $bsp_g         = 13;
 $MaxNoNodes    = 1900;
 
-$icondir    = ( $INSTALLING ? "$libdir/icons" 
+$icondir    = ( $INSTALLING ? "$datadir/icons" 
 	                    : "$FPTOOLS_TOP_ABS/ghc/utils/prof/icons" );
  
 $xmlparser  = ( $INSTALLING ? "$libexecdir/xmlparser"
diff -urN ghc-6.2.1.org/ghc/utils/prof/icons/Makefile ghc-6.2.1/ghc/utils/prof/icons/Makefile
--- ghc-6.2.1.org/ghc/utils/prof/icons/Makefile	2004-06-10 11:22:04.797016320 +0200
+++ ghc-6.2.1/ghc/utils/prof/icons/Makefile	2004-06-10 11:22:22.461330936 +0200
@@ -7,7 +7,7 @@
 TOP=../../..
 include $(TOP)/mk/boilerplate.mk
 
-override datadir=$(libdir)/icons
+override datadir:=$(datadir)/icons
 INSTALL_DATAS=$(wildcard *.xbm)
 
 include $(TOP)/mk/target.mk
diff -urN ghc-6.2.1.org/ghc/utils/prof/Makefile ghc-6.2.1/ghc/utils/prof/Makefile
--- ghc-6.2.1.org/ghc/utils/prof/Makefile	2004-06-10 11:22:04.805015104 +0200
+++ ghc-6.2.1/ghc/utils/prof/Makefile	2004-06-10 11:22:22.462330784 +0200
@@ -29,7 +29,7 @@
 endif
 
 ifneq "$(BIN_DIST)" "1"
-SCRIPT_SUBST_VARS += libdir libexecdir
+SCRIPT_SUBST_VARS += datadir libdir libexecdir
 endif
 
 # don't recurse on 'make install'
diff -urN ghc-6.2.1.org/libraries/Makefile ghc-6.2.1/libraries/Makefile
--- ghc-6.2.1.org/libraries/Makefile	2004-06-10 11:22:05.612892288 +0200
+++ ghc-6.2.1/libraries/Makefile	2004-06-10 11:23:40.178516128 +0200
@@ -77,3 +77,9 @@
 	done
 
 endif # NO_HADDOCK_DOCS
+
+CLEAN_FILES += config.status libraries.txt
+
+extraclean ::
+	$(RM) -rf $(HTML_DIR)
+
diff -urN ghc-6.2.1.org/libraries/OpenGL/specs/enumerant/Makefile ghc-6.2.1/libraries/OpenGL/specs/enumerant/Makefile
--- ghc-6.2.1.org/libraries/OpenGL/specs/enumerant/Makefile	2004-06-10 11:22:05.320936672 +0200
+++ ghc-6.2.1/libraries/OpenGL/specs/enumerant/Makefile	2004-06-10 11:22:22.465330328 +0200
@@ -5,6 +5,7 @@
 
 # -----------------------------------------------------------------------------
 
+HC           = $(GHC_INPLACE)
 MKDEPENDHS   = $(GHC_INPLACE)
 HS_PROG      = ConvertEnumSpec
 CLEAN_FILES += Main.hi test.stdout.tmp
@@ -13,9 +14,9 @@
 
 ghc_ge_504 = $(shell if (test $(GhcCanonVersion) -ge 504); then echo YES; else echo NO; fi)
 
-ifeq "$(ghc_ge_504)" "NO"
-SRC_HC_OPTS += -package data -package text
-endif
+#ifeq "$(ghc_ge_504)" "NO"
+#SRC_HC_OPTS += -package data -package text
+#endif
 
 test: ConvertEnumSpec
 	$(RM) test.stdout.tmp
