--- ghc-6.12.1/configure.ac~	2009-12-10 19:11:33.000000000 +0100
+++ ghc-6.12.1/configure.ac	2010-04-08 15:50:31.639124660 +0200
@@ -274,7 +274,7 @@
 
 checkArch() {
     case $1 in
-    alpha|arm|hppa|hppa1_1|i386|ia64|m68k|mips|mipseb|mipsel|powerpc|powerpc64|rs6000|s390|sparc|sparc64|vax|x86_64)
+    alpha|arm|hppa|hppa1_1|i386|i486|i586|i686|ia64|m68k|mips|mipseb|mipsel|powerpc|powerpc64|rs6000|s390|sparc|sparc64|vax|x86_64)
         ;;
     *)
         echo "Unknown arch $1"
@@ -285,7 +285,7 @@
 
 checkVendor() {
     case $1 in
-    dec|unknown|hp|apple|next|sun|sgi|ibm)
+    pld|dec|unknown|hp|apple|next|sun|sgi|ibm)
         ;;
     *)
         echo "Unknown vendor $1"
--- ghc-6.12.1/libraries/terminfo/configure.ac~	2009-09-15 08:09:05.000000000 +0200
+++ ghc-6.12.1/libraries/terminfo/configure.ac	2010-04-08 16:43:45.040705349 +0200
@@ -13,7 +13,7 @@
     [curses_libraries=NONE])
 
 
-TERMINFO_INCLUDE_DIRS=
+TERMINFO_INCLUDE_DIRS=/usr/include/ncursesw
 TERMINFO_LIB_DIRS=
 if test "x$curses_libraries" != "xNONE"; then
   LDFLAGS="-L$curses_libraries $LDFLAGS"
